Mathlib/Data/Multiset/Basic.lean

```diff
@@ -2327,6 +2332,57 @@ theorem map_count_True_eq_filter_card (s : Multiset α) (p : α → Prop) [Decid
   simp only [count_eq_card_filter_eq, filter_map, card_map, Function.id_comp,
     eq_true_eq_id, Function.comp_apply]
 
+theorem sub_singleton [DecidableEq α] (a : α) (s : Multiset α) : s - {a} = s.erase a := by
+  ext
+  simp [Multiset.erase_singleton, Multiset.count_singleton]
+  split <;> simp_all
+
+theorem singleton_add_sub_of_cons_add [DecidableEq α] {a a0: α} {M X : Multiset α}
+    (H : a ::ₘ M = X + {a0}) : M = X + {a0} - {a} := by
+  by_cases hyp : a = a0
+  · rw [hyp, add_comm] at H
+    simp_all [Multiset.singleton_add]
+  · have a0_a: a0 ≠ a := by rw [eq_comm] at hyp; exact hyp
+    ext b
+    simp [Multiset.count_cons, Multiset.count_singleton, Multiset.count_add]
+    have H : Multiset.count b (a ::ₘ M) = Multiset.count b (X + {a0}) := by simp_all only
+    [Multiset.count_add]
+    by_cases ba : b = a
+    · rw [ba] at *
+      have : (a ::ₘ M).count a = M.count a + 1 := by simp
+      simp_all
+    by_cases ba0 : b = a0
+    · have : (a ::ₘ M).count a0 = X.count a0 + 1 := by
+        subst_eqs
+        rw [add_comm, Multiset.singleton_add] at H
+        simp_all
+      have : M.count a0 = Multiset.count a0 (a ::ₘ M) := by
+        have : a0 ≠ a := by simp_all
+        rw [Multiset.count_cons_of_ne this M]
+      simp_all
+    · have : M.count b = (a ::ₘ M).count b := by
+        have : b ≠ a := by simp_all
+        rw [Multiset.count_cons_of_ne this M]
+      rw [this]
+      simp_all
+
+theorem mem_sub_of_not_mem_of_mem {α} [DecidableEq α] (x : α) (X Y: Multiset α) (x_in_X : x ∈ X)
+    (x_notin_Y : x ∉ Y) : x ∈ X - Y  := by
+  have : Multiset.count x X ≥ 1 := by
+    rw [← Multiset.one_le_count_iff_mem] at x_in_X
+    exact x_in_X
+  rw [← Multiset.one_le_count_iff_mem, Multiset.count_sub]
+  simp_all
+
+theorem inter_add_sub_of_eq {α} [dec : DecidableEq α] {M N P Q: Multiset α} (h : M + N = P + Q) :
+    N = N ∩ Q + (P - M) := by
```

Comment:
Can you put a few more brackets so that it's clear whether it's supposed to be read as
```suggestion
    N = N ∩ (Q + (P - M)) := by
```
or
```suggestion
    N = (N ∩ Q) + (P - M) := by
```